### PR TITLE
Add DH, DSA, ECDH and ECDSA support in PKA Engine.

### PR DESCRIPTION
**Pull request by mahantesh-nvidia**

### CODE REVIEW

File: src/engine/helper/ec_local.h [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <openssl/obj_mac.h>
+#include <openssl/ec.h>
+#include <openssl/bn.h>
+
+/* Note: Although these structures are internal to OpenSSL, they are defined
```
Comment (kblaiech) [inline src/engine/helper/ec_local.h:16]: multi-line comments should be as follow:

     /*
      * Multi-line
      * comment.
      */

File: src/engine/helper/ec_local.h [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+ *       here, because in order to override ecc point operations with pka
+ *       API's the fields inside the structure have to be changed directly as
+ *       OpenSSL doesn't provide API's to do so.
+ *
+ *       Care needs to be taken to preserve the order and members of the structure
```
Comment (kblaiech) [inline src/engine/helper/ec_local.h:21]: Good point! We have to keep it consistent across supported OpenSSL libraries

File: src/engine/helper/ec_local.h [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+ *       same as in the respective versions of OpenSSL.
+ *
+ *       Structures are mirrored from crypto/ec/ec_local.h inside openssl github repo.
+ *       Prior to openssl 1.1.1 version, this file is named as ec_lcl.h
+ */
```
Comment (kblaiech) [inline src/engine/helper/ec_local.h:26]: I think this not should be at the top, prior to the copyright:

  /*
   * Note goes here.
   *
   * Copyright goes here.
   *
   * License goes here.
   */

File: src/engine/helper/ec_local.h [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+/* Note: Although these structures are internal to OpenSSL, they are defined
+ *       here, because in order to override ecc point operations with pka
+ *       API's the fields inside the structure have to be changed directly as
+ *       OpenSSL doesn't provide API's to do so.
+ *
+ *       Care needs to be taken to preserve the order and members of the structure
+ *       same as in the respective versions of OpenSSL.
+ *
+ *       Structures are mirrored from crypto/ec/ec_local.h inside openssl github repo.
+ *       Prior to openssl 1.1.1 version, this file is named as ec_lcl.h
```
Comment (kblaiech) [inline src/engine/helper/ec_local.h:25]: Is it an exact copy of the file ec_lcl.h, or only used structures are copied?

File: src/engine/helper/pka_helper.c [hunk 5]
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
...
+    // Now init the operand buffer
+    operand->buf_ptr    = malloc(buf_max_len);
+    operand->buf_len    = buf_max_len;
+    memset(operand->buf_ptr, 0, buf_max_len);
+
+    // Now fill the operand buf.
+    operand_byte_copy(operand, (uint8_t *) bn_buf_ptr, buf_len);
+
+}
+static void bignum_to_ecc_operand(pka_bignum_t  *bignum,
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:210]: Please leave an empty line

File: src/engine/helper/pka_helper.c [hunk 6]
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
...
+                                  pka_operand_t *operand)
+{
+    if (operand == NULL)
+        return;
+
+    uint32_t byte_len, byte_max_len;
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:216]: Must be at the top. Please do not mix code and variable declaration.

File: src/engine/helper/pka_helper.c [hunk 7]
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
     free(operand);
 }
 
+/* ECC related helpers */
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:184]: coding style issue: I think in this file comments are as follow:

  // One line comment

  //
  // Multi-line comment
  //

File: src/engine/helper/pka_helper.c [hunk 8]
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
...
+    {
+        byte_len     = bignum->top  * PKA_BYTES;
+        byte_max_len = bignum->dmax * PKA_BYTES;
+
+        make_ecc_operand(bignum->d, byte_len, byte_max_len, 0, operand);
+    }
+    return;
+}
+
+static void malloc_ecc_operand(uint32_t       buf_len,
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:228]: Can't we use malloc_operand() as defined in pka_helper.c?
I think this is redundant code.
same for the functions below?

File: src/engine/helper/pka_helper.c [hunk 9]
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
...
+    point = malloc(sizeof(ecc_point_t));
+    memset(point, 0, sizeof(ecc_point_t));
+
+    malloc_ecc_operand(x_buf_len, &point->x);
+    malloc_ecc_operand(y_buf_len, &point->y);
+
+    return point;
+}
+
+static void free_ecc_operand(pka_operand_t *operand)
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:257]: same here. I think we can re-use free_operand() function instead

File: src/engine/helper/pka_helper.c [hunk 10]
```diff
@@ -181,6 +181,103 @@ static void free_operand(pka_operand_t *operand)
...
+{
+    uint8_t *buf_ptr;
+
+    if (operand == NULL)
+        return;
+
+    buf_ptr = operand->buf_ptr;
+
+    if (buf_ptr != NULL)
+        free(buf_ptr);
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:267]: free(operand)?

File: src/engine/helper/pka_helper.c [hunk 11]
```diff
@@ -629,6 +837,143 @@ int pka_rsa_mod_exp_crt(pka_bignum_t  *bn_value,
...
+    return_if_handle_invalid(tls_handle);
+
+    bignum_to_ecc_operand(bn_p, &(curve.p));
+    bignum_to_ecc_operand(bn_a, &(curve.a));
+    bignum_to_ecc_operand(bn_b, &(curve.b));
+    bignum_to_ecc_operand(bn_x, &(pt.x));
+    bignum_to_ecc_operand(bn_y, &(pt.y));
+    mltpr  = bignum_to_operand(bn_multiplier);
+
+    result = pka_do_ecc_pt_mult(tls_handle, &curve, &pt, mltpr);
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.c:872]: To be consistent with the rest of the function calls, I think it is better to declare pka_do_ecc_pt_mult as:

   int pka_do_ecc_pt_mult(..., ecc_point *result)

File: src/engine/helper/pka_helper.h [hunk 12]
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
                         pka_bignum_t *bn_qinv,
                         pka_bignum_t *bn_result);
 
+// This functions implements the elliptic curve point addition using
+// Bluefield PKA hardware.
+int pka_ecc_bn_pt_add(pka_bignum_t *bn_p,
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.h:117]: pka_bn_ecc_pt_add() to be consistent with the above.

File: src/engine/helper/pka_helper.h [hunk 13]
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
...
+                      pka_bignum_t *bn_x1,
+                      pka_bignum_t *bn_y1,
+                      pka_bignum_t *bn_x2,
+                      pka_bignum_t *bn_y2,
+                      pka_bignum_t *bn_result_x,
+                      pka_bignum_t *bn_result_y);
+
+// This functions implements the elliptic curve point multiplication using
+// Bluefield PKA hardware.
+int pka_ecc_bn_pt_mult(pka_bignum_t *bn_p,
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.h:129]: Same here

File: src/engine/helper/pka_helper.h [hunk 14]
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
...
+                      pka_bignum_t *bn_a,
+                      pka_bignum_t *bn_b,
+                      pka_bignum_t *bn_x1,
+                      pka_bignum_t *bn_y1,
+                      pka_bignum_t *bn_x2,
+                      pka_bignum_t *bn_y2,
+                      pka_bignum_t *bn_result_x,
+                      pka_bignum_t *bn_result_y);
+
+// This functions implements the elliptic curve point multiplication using
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.h:127]: typo. function instead of functions

File: src/engine/helper/pka_helper.h [hunk 15]
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
                         pka_bignum_t *bn_qinv,
                         pka_bignum_t *bn_result);
 
+// This functions implements the elliptic curve point addition using
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.h:115]: typo. function instead of functions

File: src/engine/helper/pka_helper.h [hunk 16]
```diff
@@ -112,6 +112,35 @@ int pka_rsa_mod_exp_crt(pka_bignum_t *bn_value,
...
+                       pka_bignum_t *bn_b,
+                       pka_bignum_t *bn_x,
+                       pka_bignum_t *bn_y,
+                       pka_bignum_t *bn_multiplier,
+                       pka_bignum_t *bn_result_x,
+                       pka_bignum_t *bn_result_y);
+
+// This function implements the modular inverse using BlueField
+// PKA hardware.
+int pka_mod_inv(pka_bignum_t *bn_value,
```
Comment (kblaiech) [inline src/engine/helper/pka_helper.h:140]: pka_bn_mod_inv?

File: src/lib/pka_queue.c [hunk 17]
```diff
@@ -155,6 +155,33 @@ static uint16_t pka_concat6_wlen(uint32_t word_len,
     return total_wlen;
 }
 
+// Adjusts the value present in the array pointed by buf_ptr.
+// Adjustment is done by removing zero padding at the most significant bytes.
+// Once this is done, align the new length around 8-byte boundary.
+// Note: Here it's assumed that the value is in Little Endian format.
+static void pka_adjust_rslt(pka_operand_t *result)
```
Comment (kblaiech) [inline src/lib/pka_queue.c:162]: I believe "adjust" is a bit confusing. Could you please call it something like: pka_queue_no_zero_pad_rslt or any other name that explicitly indicates the type of the operation?

Note that, by convention, all functions in pka_queue.c have the prefix pka_queue_*

File: src/lib/pka_queue.c [hunk 18]
```diff
@@ -155,6 +155,33 @@ static uint16_t pka_concat6_wlen(uint32_t word_len,
...
+    // First find the most significant byte based upon the len, and
+    // then move backwards over all zero bytes.
+    uint8_t  *byte_ptr;
+    uint32_t  result2_len;
+
+    result2_len = result->actual_len;
+    byte_ptr    = result->buf_ptr;
+
+    // Result from pka hardware is always in Little Endian.
+    byte_ptr = &byte_ptr[result2_len - 1];
```
Comment (kblaiech) [inline src/lib/pka_queue.c:173]: This is true. But is it always true? what if the hardware change? Can we consider an if statement that handles both BE and LE. I think the byte order is stored somewhere, thus we can easily retrieve the information and apply the right logic.

File: src/engine/helper/ec_local.h (hunk 1 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/ec_local.h:16]: Yes, Thanks.

File: src/engine/helper/ec_local.h (hunk 4 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/ec_local.h:25]: Only the used structures are copied. There are many other definitions which require inclusion of other internal headers. These other definitions are not useful to us. Hence, only used structures are copied.

File: src/engine/helper/ec_local.h (hunk 3 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/ec_local.h:26]: Ack, Thanks.

File: src/engine/helper/pka_helper.c (hunk 7 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.c:184]: Ack, Thanks.

File: src/engine/helper/pka_helper.c (hunk 5 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.c:210]: Will do, Thanks.

File: src/engine/helper/pka_helper.c (hunk 6 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.c:216]: Ack, Thanks.

File: src/engine/helper/pka_helper.c (hunk 8 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.c:228]: As `ecc_point_t` has x, y variables of type `pka_operand_t` and not `pka_operand_t*`, we don't dynamically allocate and hence free this memory. As this `malloc()` is done for operand in` malloc_operand()`, we cannot use the same here. As discussed, I'll add "_buf" suffix to the function name as the current name is misleading.

File: src/engine/helper/pka_helper.c (hunk 11 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.c:872]: As discussed, the current declaration is consistent with the rest of function calls. But I'll initialize the `rc` variable at the declaration time.

File: src/engine/helper/pka_helper.h (hunk 12 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.h:117]: Ack, Thanks.

File: src/engine/helper/pka_helper.h (hunk 14 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.h:127]: Ack, Thanks.

File: src/engine/helper/pka_helper.h (hunk 13 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.h:129]: Ack, Thanks.

File: src/lib/pka_queue.c (hunk 17 above)
Comment (mahantesh-nvidia) [reply src/lib/pka_queue.c:162]: Yes, I'll replace "adjust" with something more relevant. However, as discussed, the inclusion of "queue" in the function name is not necessary as this is a static function. Thanks

File: src/lib/pka_queue.c (hunk 18 above)
Comment (mahantesh-nvidia) [reply src/lib/pka_queue.c:173]: I'll add the BE logic, Thanks.

File: src/engine/helper/pka_helper.h (hunk 16 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.h:140]: Yes, correct. Thanks.

File: src/engine/helper/pka_helper.c (hunk 10 above)
Comment (mahantesh-nvidia) [reply src/engine/helper/pka_helper.c:268]: `ecc_point_t `has x,y variables declared as `pka_operand_t`, hence these operands are not dynamically created.

**Comment by mahantesh-nvidia** [comment]
> Looks good to me except few comments I would like you to address and other minor issues related to the coding style.

Thanks for the review @kblaiech , I have addressed your review comments.

**Pull request merged**